File: pulse_mobile/lib/model/bicycle/bicycle.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pulse_mobile/model/available_size/available_size.dart';
import 'package:pulse_mobile/model/product_image/product_image.dart';

import '../brand/brand.dart';
import '../abstract/product.dart';
import '../product_category/product_category.dart';

part 'bicycle.g.dart';

@JsonSerializable()
class Bicycle extends Product {
  double? weight;
  int? productionYear;
  String? wheelSize;
  List<AvailableSize>? availableSizes;

  Bicycle();

  factory Bicycle.fromJson(Map<String, dynamic> json) =>
      _$BicycleFromJson(json);

  Map<String, dynamic> toJson() => _$BicycleToJson(this);

  bool isAvailable() {
    return (availableSizes
                ?.map((e) => e.availableQty ?? 0)
                .reduce((a, b) => a + b) ??
            0) >
        0;
  }
}

File: PULSE.Model/Bicycle.cs
namespace PULSE.Model
{
    public class Bicycle : Product
    {
        public decimal? Weight { get; set; }
        public int? ProductionYear { get; set; }
        public string? WheelSize { get; set; }

        public string formattedSizes => string.Join(", ", AvailableSizes.Select(s => s.BicycleSize));

        public virtual ProductCategory? ProductCategory { get; set; }
        public virtual ICollection<AvailableSize> AvailableSizes { get; set; }
    }
}
